File: src/lib/planRuns.ts
import { invoke } from "@tauri-apps/api/core";
import { listen } from "@tauri-apps/api/event";
import type { UnlistenFn } from "@tauri-apps/api/event";

export type PlanRunStatus =
  | "pending"
  | "running"
  | "succeeded"
  | "failed"
  | "cancelled"
  | "paused"
  | "awaiting_review";

export type RunnerKind = "native" | "omp";

export type PlanQueueEntry = {
  id: string;
  sessionId: string;
  planId: string;
  sortOrder: number;
  createdAt: number;
};

export type ExecutionProfile = {
  concurrency: number;
  providerId: string;
  modelId: string;
  effortLevel?: string;
};

export type PlanOverride = {
  planId: string;
  providerId: string;
  modelId: string;
  effortLevel?: string;
};

export type PlanRun = {
  id: string;
  planId: string;
  sessionId: string;
  chatSessionId?: string;
  workspacePath?: string;
  status: PlanRunStatus;
  runnerKind: RunnerKind;
  error?: string;
  stepsOutput: PlanRunStepOutput[];
  startedAt?: number;
  finishedAt?: number;
  createdAt: number;
};

export type PlanRunStepOutput = {
  stepId: string;
  kind: string;
  status: string;
  output?: string;
  error?: string;
};

export type PlanRunEvent = {
  runId: string;
  sessionId: string;
  planId: string;
  status: PlanRunStatus;
  chatSessionId?: string;
  error?: string;
};

export type EnqueuePlanRequest = {
  sessionId: string;
  planId: string;
};

export type StartQueueRequest = {
  sessionId: string;
  profile: ExecutionProfile;
  planOverrides?: PlanOverride[];
};

export async function enqueuePlan(
  request: EnqueuePlanRequest,
): Promise<PlanQueueEntry> {
  return invoke<PlanQueueEntry>("plan_run_enqueue", { request });
}

export async function listPlanQueue(
  sessionId: string,
): Promise<PlanQueueEntry[]> {
  return invoke<PlanQueueEntry[]>("plan_run_list_queue", { sessionId });
}

export async function reorderPlanRun(
  sessionId: string,
  entryId: string,
  newOrder: number,
): Promise<void> {
  return invoke<void>("plan_run_reorder", { sessionId, entryId, newOrder });
}

export async function removePlanRun(entryId: string): Promise<void> {
  return invoke<void>("plan_run_remove", { entryId });
}

export async function startQueue(request: StartQueueRequest): Promise<void> {
  return invoke<void>("plan_run_start", { request });
}

export async function pauseQueue(sessionId: string): Promise<void> {
  return invoke<void>("plan_run_pause", { sessionId });
}

export async function startOmpPlanRun(
  sessionId: string,
  planId: string,
): Promise<PlanRun> {
  return invoke<PlanRun>("plan_run_start_omp", { sessionId, planId });
}

export async function assignPlanToChat(
  planId: string,
  chatSessionId: string,
): Promise<PlanRun> {
  return invoke<PlanRun>("plan_assign_to_chat", { planId, chatSessionId });
}

export async function cancelPlanRun(
  runId: string,
  cancelPlan: boolean,
): Promise<void> {
  return invoke<void>("plan_run_cancel", { runId, cancelPlan });
}

export async function completePlanRun(
  runId: string,
  succeeded: boolean,
): Promise<void> {
  return invoke<void>("plan_run_complete", { runId, succeeded });
}

export async function checkPlanRunCompletion(
  runId: string,
): Promise<[number, number]> {
  return invoke<[number, number]>("plan_run_check_completion", { runId });
}


export async function listPlanRuns(sessionId: string): Promise<PlanRun[]> {
  return invoke<PlanRun[]>("plan_run_list", { sessionId });
}

export async function getPlanRun(runId: string): Promise<PlanRun | null> {
  return invoke<PlanRun | null>("plan_run_get", { runId });
}

/// Listen for plan_run:// events. Returns an unlisten function.
export function onPlanRunEvent(
  cb: (event: PlanRunEvent) => void,
): Promise<UnlistenFn> {
  return listen<PlanRunEvent>("plan_run://event", (e) => cb(e.payload));
}

export async function markPlanRunComplete(runId: string): Promise<void> {
  await invoke<void>("plan_run_mark_complete", { runId });
}

export type FinishOutcome = {
  runId: string;
  policy: string;
  commitSha: string | null;
  prUrl: string | null;
  mergeReady: boolean;
  error: string | null;
};

export type FinishResult =
  | { kind: "hold" }
  | { kind: "fallback_hold"; message: string }
  | { kind: "applied"; outcome: FinishOutcome };

/** Read the finish outcome persisted at run completion. Read-only — the
 *  policy itself is applied exactly once by the backend `complete_run`. */
export async function getFinishOutcome(runId: string): Promise<FinishResult> {
  const raw = await invoke<Record<string, unknown>>("plan_run_finish_outcome", { runId });
  if (raw.kind === "hold") return { kind: "hold" };
  if (raw.kind === "fallback_hold") return { kind: "fallback_hold", message: String(raw.message ?? "") };
  if (raw.kind === "applied") return { kind: "applied", outcome: raw.outcome as FinishOutcome };
  return { kind: "hold" };
}

File: src-tauri/src/commands/plan_runs.rs
use tauri::AppHandle;

use crate::{
    models::plan_run::{EnqueuePlanRequest, PlanQueueEntry, PlanRun, StartQueueRequest},
    services::plan_runner_service::PlanRunnerService,
};

#[tauri::command]
pub fn plan_run_enqueue(request: EnqueuePlanRequest) -> Result<PlanQueueEntry, String> {
    PlanRunnerService::enqueue(request)
}

#[tauri::command]
pub fn plan_assign_to_chat(
    app: AppHandle,
    plan_id: String,
    chat_session_id: String,
) -> Result<PlanRun, String> {
    let run = PlanRunnerService::assign_to_chat(&app, &plan_id, &chat_session_id)?;
    // Start the agent turn — assignment alone only seeds context.
    PlanRunnerService::kickoff_assigned_run(&app, &chat_session_id);
    Ok(run)
}

#[tauri::command]
pub fn plan_run_list_queue(session_id: String) -> Result<Vec<PlanQueueEntry>, String> {
    PlanRunnerService::list_queue(&session_id)
}

#[tauri::command]
pub fn plan_run_reorder(
    session_id: String,
    entry_id: String,
    new_order: i64,
) -> Result<(), String> {
    PlanRunnerService::reorder(&session_id, &entry_id, new_order)
}

#[tauri::command]
pub fn plan_run_remove(entry_id: String) -> Result<(), String> {
    PlanRunnerService::remove_from_queue(&entry_id)
}

#[tauri::command]
pub fn plan_run_start(app: AppHandle, request: StartQueueRequest) -> Result<(), String> {
    PlanRunnerService::start_queue(app, request)
}

#[tauri::command]
pub fn plan_run_pause(session_id: String) -> Result<(), String> {
    PlanRunnerService::pause_queue(&session_id)
}

#[tauri::command]
pub fn plan_run_cancel(app: AppHandle, run_id: String, cancel_plan: bool) -> Result<(), String> {
    PlanRunnerService::cancel_run(&app, &run_id, cancel_plan)
}

#[tauri::command]
pub fn plan_run_complete(app: AppHandle, run_id: String, succeeded: bool) -> Result<(), String> {
    PlanRunnerService::complete_run(&app, &run_id, succeeded)
}

#[tauri::command]
pub fn plan_run_mark_complete(app: AppHandle, run_id: String) -> Result<(), String> {
    PlanRunnerService::mark_complete(&app, &run_id)
}

#[tauri::command]
pub fn plan_run_finish_outcome(run_id: String) -> Result<serde_json::Value, String> {
    // Read-only: returns the outcome persisted by `complete_run`. NEVER
    // re-executes the policy (no commits, pushes, PRs, or queue writes).
    Ok(PlanRunnerService::get_finish_outcome(&run_id)?
        .unwrap_or_else(|| serde_json::json!({ "kind": "hold" })))
}

#[tauri::command]
pub fn plan_run_check_completion(app: AppHandle, run_id: String) -> Result<(u32, u32), String> {
    PlanRunnerService::check_run_completion(&app, &run_id)
}

#[tauri::command]
pub fn plan_run_start_omp(
    app: AppHandle,
    session_id: String,
    plan_id: String,
) -> Result<PlanRun, String> {
    PlanRunnerService::start_omp_run(&app, &session_id, &plan_id)
}

#[tauri::command]
pub fn plan_run_list(session_id: String) -> Result<Vec<PlanRun>, String> {
    PlanRunnerService::list_runs(&session_id)
}

#[tauri::command]
pub fn plan_run_list_by_project(project_path: String) -> Result<Vec<PlanRun>, String> {
    PlanRunnerService::list_runs_for_project(&project_path)
}

#[tauri::command]
pub fn plan_run_list_by_plan(plan_id: String) -> Result<Vec<PlanRun>, String> {
    PlanRunnerService::list_runs_by_plan(&plan_id)
}

#[tauri::command]
pub fn plan_run_get(run_id: String) -> Result<Option<PlanRun>, String> {
    PlanRunnerService::get_run(&run_id)
}
